File: Python/DMOJ/hidenseek.py
from itertools import permutations as p
from copy import deepcopy
def main(n, m, t, start, hiders, arr):
    #!Creating permutations doesn't change but the idea that mix up indices to save space makes sense. 
    #*The permutation method really runs out of memory too. 
    #!The iterable cannot be hiders since 
    temp = []
    for i in range(t):
        temp.append(i)
    permutations = list(p(temp))
    #!But if this is the case, I need to make sure the front element actually comes at front. 
    '''
    queue = [[0]]

    while (len(queue) > 0):
        curr = queue.pop(0)
        if (len(curr) == t+1):
            permutations.append(curr)
        else:
            for i in range(len(hiders)):
                if i not in curr:
                    #*I think it is the array copying that really takes the toll. 
                    temp = curr.copy()
                    temp.append(i)
                    queue.append(temp)
    '''
    
    dp = {}
    queue = []
    result = 99999
    visited = deepcopy(arr)

    #*First have to do it separately for the first element. 
    


    #!Really cannot spend entire memory tupling stuff. 
    for i in range(len(permutations)):
        #*No issue for j+1 exceeding because since there must be a start point and at least one hider, no permutation will have length 2 or smaller. 
        curr_length = 0
        #!Since we aren't including the first element, the loop counter must change. 
        for j in range(t):
            #*Changing the visited array. Only use little extra memory. 
            for row in range(n):
                for col in range(m):
                    visited[row][col] = arr[row][col]
            #!Just need to adjust the hashability types. 
            if (tuple(hiders[permutations[i][j]]), tuple(hiders[permutations[i][j+1]])) in dp:
                curr_length += dp[tuple(hiders[permutations[i][j]], tuple(hiders[permutations[i][j+1]]))]
                break
            elif (tuple(hiders[permutations[i][j+1]]), tuple(hiders[permutations[i][j]])) in dp:
                curr_length += dp[tuple(hiders[permutations[i][j+1]]), tuple(hiders[permutations[i][j]])]
                break
            else:
                #*Last item to count how the distance. 
                #!Need to check the indexing a little bit. Since it has to be hiders[][0]
                #!Still 4. 
                queue = [[hiders[permutations[i][j]][0], hiders[permutations[i][j]][1], 0]]
                while len(queue) > 0:
                    curr = queue.pop(0)
                    if (curr[0]==hiders[permutations[i][j+1]][0] and curr[1]==hiders[permutations[i][j+1]][1]):
                        curr_length += curr[2]
                        dp[(tuple(hiders[permutations[i][j]]), tuple(hiders[permutations[i][j+1]]))] = curr[2]
                        break
                    #*Only checking visited also checks arr. 
                    elif 0 <= curr[0] < n and 0 <= curr[1] < m and visited[curr[0]][curr[1]] != 'X':
                        visited[curr[0]][curr[1]] = 'X'
                        queue.append([curr[0]+1, curr[1], curr[2]+1])
                        queue.append([curr[0]-1, curr[1], curr[2]+1])
                        queue.append([curr[0], curr[1]+1, curr[2]+1])
                        queue.append([curr[0], curr[1]-1, curr[2]+1])
                    
        result = min(result, curr_length)
    return result
    

n, m, t = input().strip().split(" ")
n, m, t = int(n), int(m), int(t)

arr = []
hiders = []
for i in range(n):
    s = input()
    temp = []
    for j in range(m):
        temp.append(s[j])
        #!I guess I could insert it to the beginning to ensure starting always shows up in the beginning.
        if s[j] == 'G':
            start = [i, j]
            #hiders.insert(0, [i, j])
        elif s[j] == 'H':
            hiders.append([i, j])
    arr.append(temp)

print(main(n, m, t, start, hiders, arr))



#*For Python, memory of creating all these permutations in row, col form is what causes the memory error. 
#*For that, I honestly think I need to first find out the individual lengths, store it in maybe a dictionary, then run a loop, storing the current length in an int using BFS. 
'''
def main(n, m, t, start, hiders, arr):
    d = {}
    visited = deepcopy(arr)
    #*Find the distance between hiders. 
    for i in range(len(hiders)):
        #*Don't run the reverse.
        for j in range(i+1, len(hiders)):
            
            for row in range(n):
                for col in range(m):
                    visited[row][col] = arr[row][col]
            queue = [(hiders[i][0], hiders[i][1], 0)]
            while len(queue) > 0:
                curr = queue.pop(0)
                if (curr[0] == hiders[j][0] and curr[1] == hiders[j][1]):
                    d[(hiders[i], hiders[j])] = curr[2]
                    break
                else:
                    queue.append((curr[0]+1, curr[1], curr[2]+1))
                    queue.append((curr[0]-1, curr[1], curr[2]+1))
                    queue.append((curr[0], curr[1]+1, curr[2]+1))
                    queue.append((curr[0], curr[1]-1, curr[2]+1))
    
    #*Starting queue. 
    queue = [start]
    result = 9999999
    
    while len(queue) > 0:
        #?How am I supposed to know which paths I have went to and which I haven't?
        #*Ultimately, storing all the possibilities costs me a lot of memory. 
        #*Travelling salesman problem in linear time.


    #*Just learn from bobhob's submission. https://dmoj.ca/src/88584
    #!Append all hiders into an array. 

    #*create a 6 by 5 matrix for what?
    #*Then create a visited matrix (wouldn't it be easier with a set?)
    #!It just seems like I am going to wherever I can go to. 
    #*Creative data structure. Instead of using queue to store every where visited, use array to know. 

    #*First go from griffy and check whether they can be visited?

            

n, m, t = input().strip().split(" ")
n, m, t = int(n), int(m), int(t)

arr = []
hiders = []
for i in range(n):
    s = input()
    temp = []
    for j in range(m):
        temp.append(s[j])

        if s[j] == 'H':
            hiders.append((i, j))
        if s[j] == 'G':
            start = (i, j)
            #!Technically, I could treat the starting point as one of the hiders. 
            hiders.append((i, j))
    arr.append(temp)

print(main(n, m, t, start, tuple(hiders), arr))
'''

File: C++/DMOJ cpp/hidenseek.cpp
#include <iostream>
#include <vector>
#include <set>
#include <cmath>
#include <cstdlib>
#include <stdio.h>
#include <assert.h>
#include <algorithm>
#include <map>
using namespace std;


int showVector(vector <vector <vector <int>>> arr){
    for (int i=0;i<arr.size();i++){
        for (int j=0;j<arr[i].size();j++){
            for (int k=0;k<arr[i][j].size();k++){
                cout << arr[i][j][k];
            }
            cout << " ";
        }
        cout << endl;
    }
    return 1;
}

int hidenseek(int n, int m, int t, vector <int> start, vector <vector <int>> hiders, vector <vector <char>> arr){
    //*First generate all possible orders. 
    vector <vector <int>> curr = {start};
    vector <vector <vector <int>>> queue = {curr};
    vector <vector <vector <int>>> permutations;
    vector <vector <int>> temp;
    //!Store all the points I have added. 
    //vector <
    //!For some reason, number of permutations isn't increasing. 
    while (queue.size() > 0){
        curr = queue[0];
        queue.erase(queue.begin());

        //!If we have already chosen all, I don't need to do anymore stuff. 
        if (curr.size() == t+1){
            permutations.push_back(curr);
        }
        //*Need to make sure it is in range. 
        //!I see, because everything was registered based on array indices, I need to follow that.
        //!Actually, I don't have to check for range since everything is in range.  
        else{
            //*Iterate through all hiders and check whether they already exist. 
            for (int i=0; i<hiders.size();i++){
                //*Find function has some kind of error. 
                //if (find(curr.begin(), curr.end(), hiders[i]) == curr.end()){
                //!
                if (count(curr.begin(), curr.end(), hiders[i])==0){
                    //*Element doesn't exist. 
                    temp = curr;
                    temp.push_back(hiders[i]);
                    queue.push_back(temp);
                }
            }
        }
    }
    //!Segmentation fault is happening in the showing loop. 
    //showVector(permutations);

    //return 1;

    //*Loop through all the permutations. 
    //*Loop through every step
    //*Find the shortest path using BFS. 
    //*Probably store in a DP array for future use. 
    //!I have to map the starting point to ending point. Doesn't matter if it is in opposite order. 
    map <vector <vector <int>>, int> subtask;
    vector <vector <int>> queue2;
    vector <vector <int>> front;
    vector <vector <int>> reversed;
    vector <int> current;
    set <vector <int>> visited;
    //!Since BFS, just keep a set for each since even if I reach it later, it doesn't work.
    int curr_length;
    int result = 999999;
    for (int i=0; i<permutations.size();i++){
        //*size()-1 because we are checking j and j+1
        curr_length = 0;
        //!Don't know why the loop is running for so long. 
        for (int j=0; j<permutations[i].size()-1;j++){
            //*Before adding, check whether it exists in queue or not. 
            visited = {};
            front = {permutations[i][j+1], permutations[i][j]};
            reversed = {permutations[i][j], permutations[i][j+1]};

            if (subtask.count(front) == 1){
                curr_length += subtask[front];
            }
            else if (subtask.count(reversed) == 1){
                curr_length += subtask[reversed];
            }
            else{
                //!Current actually has to hold the number it took to get there as well. 
                //!Convenient that I can just build an array like that and don't need to make temp arrays for everything. 
                queue2 = {{permutations[i][j][0], permutations[i][j][1], 0}};
                while (queue2.size() > 0){
                    current = queue2[0];
                    queue2.erase(queue2.begin());
                    //!There cannot be a case where each permutation has length 1 or less since it includes the starting point. 
                    bool one = current[0] == permutations[i][j+1][0];
                    bool two = current[1]==permutations[i][j+1][1];
                    if (one && two){
                        curr_length += current[2];
                        subtask[{permutations[i][j], permutations[i][j+1]}] = current[2];
                        break;
                    }

                    //!If runs out of time, I can try marking already visited nodes but won't worry for the time-being. 
                    //*Just check for out of bounds and whether there is no wall. 
                    else if (
                        0 <= current[0] &&
                        current[0] < n &&
                        0 <= current[1] &&
                        current[1] < m && 
                        arr[current[0]][current[1]] != 'X' &&
                        visited.count(current) == 0
                    ){
                        visited.insert(current);
                        queue2.push_back({current[0]+1, current[1], current[2]+1});
                        queue2.push_back({current[0]-1, current[1], current[2]+1});
                        queue2.push_back({current[0], current[1]+1, current[2]+1});
                        queue2.push_back({current[0], current[1]-1, current[2]+1});
                    }
                    //cout << "stuck";
                }
            }
        }
        result = min(result, curr_length);
    }

    return result;

}


int main(){
    //*Naive BFS won't work because there are a lot of hiders so I have to do this separately. 
    //*If I know the end nodes and all the paths I have visited so far, I think I could do a double BFS. 
    //*First for all the paths I need to check so far and second for the closest path from that point to next. 

    //?Actually, I think I could generate all the permutations first then find the shortest distance for individual edges. 

    int t, n, m;
    cin >>  n >> m >> t;

    //*Store the starting position. Store seekers position in a hashset. 
    vector <vector <char>> arr;
    vector <char> temp;
    vector <int> start;
    vector <int> temp_hider;
    vector <vector <int>> hiders;
    string s;
    for (int i=0; i<n; i++){
        temp = {};
        cin >> s;
        for (int j = 0; j < m;j++){
            temp.push_back(s[j]);
            if (s[j] == 'G'){
                start = {i, j};
            }
            else if (s[j] == 'H'){
                temp_hider = {i, j};
                hiders.push_back(temp_hider);
            }
        }
        arr.push_back(temp);
    }

    cout << hidenseek(n, m, t, start, hiders, arr);
}   